### PR TITLE
Auto-update properties when replacing a node

### LINKED ISSUE #78299: changing type of a node removes reference from `Node` properties
### Godot version

v4.1.beta.custom_build [dbd482faa]

### System information

macOS 13.3.0 - Vulkan (Forward+) - integrated Apple M1 Pro - Apple M1 Pro (8 Threads)

### Issue description

Some editor action like changing type will replace the node by another node. Right now Node properties will still point to the old Node. 

### Steps to reproduce

1. create a scene with a root node 
2. attach a script that exports a node property
3. set the property to point to the root node 
4. change the type of the root node
5. look at the property it will error out and the editor will be blank

### Minimal reproduction project

[Test.zip](https://github.com/godotengine/godot/files/11764206/Test.zip)
with MRP you can start at step 4

### PR DESCRIPTION
**Pull request by ajreckof**
Fixes #78299

The idea is to do something similar to what is done for auto updating `NodePath`

### CODE REVIEW

**Comment by Calinou** [comment]
With this PR, changing the node type from Node2D to Node does nothing – the node remains a Node2D.

Attempting to change from Node2D to Node3D prints:

```
  ./scene/main/node.cpp:1992 - Parameter "common_parent" is null.
```

While the latter case isn't expected to work, it should print an human-readable error message.

**Comment by ajreckof** [comment]
> With this PR, changing the node type from Node2D to Node does nothing – the node remains a Node2D.

I can't reproduce locally. Did you check by removing the script because right now changing the type of a node that has a script feels like it is not working because icon does not change because the node still inherit the last type and the script now dictates the icon even without `@icon`. I don't know if there is an issue tracking this.



> Attempting to change from Node2D to Node3D prints:
> 
> ```
>   ./scene/main/node.cpp:1992 - Parameter "common_parent" is null.
> ```
> 
> While the latter case isn't expected to work, it should print an human-readable error message.

I agree with you it should print a better error. I remember thinking about it but I didn't know what to do about it. I'm wondering maybe just removing the reference ?
Just for reference before this PR this was already a problem because the node would have still pointed to a Node that did not correspond to the filter.

File: editor/scene_tree_dock.cpp [hunk 1]
```diff
@@ -2493,6 +2507,79 @@ void SceneTreeDock::replace_node(Node *p_node, Node *p_by_node, bool p_keep_prop
...
+bool SceneTreeDock::_check_node_recursive(Variant &r_variant, Node *p_node, Node *p_by_node, const String type_hint) {
+	switch (r_variant.get_type()) {
+		case Variant::OBJECT: {
+			if (p_node == r_variant) {
+				if (p_by_node->is_class(type_hint) || EditorNode::get_singleton()->is_object_of_custom_type(p_by_node, type_hint)) {
+					r_variant = p_by_node;
+					return true;
+				} else {
+					r_variant = memnew(Object);
+					ERR_FAIL_V_MSG(true, vformat(TTR("The node which type was changed was stored in a variable. This variable needed a node of type %s but the new type is %s. Removing the node from this variable."), type_hint, p_by_node->get_class()));
```
Comment (ajreckof) [inline editor/scene_tree_dock.cpp:2553]: I'm not confident about my wording at all so pleas propose better wording for explaining what happens

Comment (ajreckof) [reply editor/scene_tree_dock.cpp:2553]: maybe this could be a warning instead of an error?

Comment (KoBeWi) [inline editor/scene_tree_dock.cpp:2553]:
```suggestion
					ERR_FAIL_V_MSG(true, vformat(TTR("The node's new type is incompatible with an exported variable (expected %s, but type is %s). Removing the node from this variable."), type_hint, p_by_node->get_class()));
```
I agree it should be a warning. Also you could mention the variable name if possible.

Comment (KoBeWi) [reply editor/scene_tree_dock.cpp:2553]:
```suggestion
					ERR_FAIL_V_MSG(true, vformat(TTR("The node's new type is incompatible with an exported variable (expected %s, but type is %s). Removing the node from this variable."), type_hint, p_by_node->get_class()));
```
I agree it should be a warning. Also you could mention the variable name if possible.

File: editor/scene_tree_dock.cpp [hunk 2]
```diff
@@ -2420,7 +2416,25 @@ void SceneTreeDock::_create() {
...
+	ur->add_undo_method(this, "replace_node", p_by_node, p_node, false, false);
+	ur->add_undo_reference(p_node);
+
+	perform_node_replace(nullptr, p_node, p_by_node);
+
+	ur->commit_action(false);
+}
+
+void SceneTreeDock::_replace_node(Node *p_node, Node *p_by_node, bool p_keep_properties, bool p_remove_old) {
+	ERR_FAIL_COND_MSG(!p_node->is_inside_tree(), "replace node can't be called on a node outside of tree. You might have called it twice.");
```
Comment (Calinou) [inline editor/scene_tree_dock.cpp:2437]:
```suggestion
	ERR_FAIL_COND_MSG(!p_node->is_inside_tree(), "_replace_node() can't be called on a node outside of tree. You might have called it twice.");
```

**Review by Calinou — APPROVED** [review]
Tested locally (rebased on top of `master` dcd187da9580b41688f4e5110acd767d4deab8de), it works as expected. The node type changes correctly and the exported variable reference is kept.

Code looks good to me at a glance.

**Review by KoBeWi — APPROVED** [review]

**Comment by YuriSizov** [comment]
Thanks!

**Pull request merged**